File: driver/php/Worker.php
<?php

namespace gtsd;

include("ServerPool.php");
include("Task.php");
include("Agent.php");

class Worker extends Agent
{

    protected $functions = array();
    protected $registered = false;

    public function __construct($servers)
    {
        parent::__construct($servers);

        if(!$this->serverpool->connectAll())
        {
            throw new Exception("Could not connect to any server. Stopping worker");
        }
    }

    protected function registerFunctions()
    {
        if (!$this->registered)
        {
            $this->serverpool->writeAll(new Message(Message::FUNCTION_REGISTER, array('functions' => array_keys($this->functions))));
            $this->registered = true;
        }
    }

    protected function requestWork(Message $wakeMsg = null)
    {
        $msg = new Message(Message::WORK_REQUEST);

        if($wakeMsg === null)
            $this->serverpool->writeAll($msg);
        else
        {
            $this->serverpool->write($msg, $wakeMsg->header['connection_id']);
        }
    }
    
    protected function workComplete($task)
    {
        $this->serverpool->write(new Message(Message::WORK_DONE, array('jobid' => $task->id, 'result' => $task->result)), $task->connId);
    }
    
    public function notifyWorkStatus($task)
    {
        $message = new Message(Message::WORK_STATUS, array('jobid' => $task->id, 'status' => $task->status, 'progress' => $task->progress));
        $this->serverpool->write($message, $task->connId);
    }

    public function register($function, $callback)
    {
        $this->functions[$function] = $callback;
    }

    public function work()
    {
        $this->registerFunctions();
        $this->requestWork();
        
        $message = $this->serverpool->readAll();
            
        while($this->executeMessage($message))
        {
            $message = $this->serverpool->readAll();
            
        }
        
        return true;
    }

    private function executeMessage($message)
    {
        if($message === false || !($message instanceof Message)) return true;
        
        switch ($message->getCode())
        {
            case Message::NO_WORK:
                return true;

            case Message::WAKE:
                $this->requestWork($message);
                return true;

            case Message::RUN_WORK:
                $task = $this->run($message);
                $this->workComplete($task);
                return false;
        }
        
        return true;
        
    }
    
    protected function run($message)
    {
        $task = new Task($this, $message->getBodyData('function_name'), $message->getBodyData('workload'));
        $task->updateInfo($message);
        $task->connId = $message->header['connection_id'];
        $task->status = "working";
        $callback = $this->functions[$task->function_name];

        if (is_string($callback))
            $result = call_user_func($callback, $task);
        else if (is_callable($callback, true))
        {
            if (is_array($callback))
            {
                // an array: 0 - object, 1 - method name
                list($object, $method) = $callback;
                if (is_string($object)) // static method call
                    $result = call_user_func($callback, $task);
                else if (method_exists($object, $method))
                    $result = $object->$method($task);
                else
                    throw new Exception("The $method method does not exist in ".get_class($object));
            }
            else // PHP 5.3: anonymous function
                $result = call_user_func($callback, $task);
        }

        $task->result = $result;
        return $task;
    }


}

?>

File: driver/python/worker.py
from serverpool import *
from agent import *
from message import *
from task import *
from pprint import pprint

class Worker(Agent):
	type = 'worker'
	
	def __init__(self, servers):
		Agent.__init__(self, servers)
		self.functions = {}
		self.registered = False
		self.serverpool.connectAll()
		
	def __registerFunctions(self):
		if not self.registered:
			self.serverpool.writeAll(Message(Message.FUNCTION_REGISTER, {'functions': self.functions.keys()}))
			self.registered = True
			
	def __requestWork(self, wakeMsg=None):
		msg = Message(Message.WORK_REQUEST)
		print "requesting work"
		if not isinstance(wakeMsg, Message):
			self.serverpool.writeAll(msg)
		else:
			self.serverpool.write(msg, wakeMsg.header['connection_id'])
			
		
	def __workComplete(self, task):
		#pprint((vars(task)))
		self.serverpool.write(Message(Message.WORK_DONE, {'jobid': task.id, 'result': task.result}), task.connection_id)
		
	def register(self, function, callback):
		self.functions[function] = callback
		
	def work(self):
		self.__registerFunctions()
		self.__requestWork()
		message = self.serverpool.readAll()
		
		while self.__executeMessage(message):
			message = self.serverpool.readAll()
			print "..."
		
		return True
		
	def __executeMessage(self, message):
		if not isinstance(message, Message):
			return True
			
		if message.getCode() == Message.NO_WORK:
			print "no work"
			return True
		elif message.getCode() == Message.WAKE:
			print "wake"
			self.__requestWork()
			return True
		elif message.getCode() == Message.RUN_WORK:
			print "run work"
			task = self.__run(message)
			self.__workComplete(task)
			return False
		
		return True
		
	def __run(self, m):
		task = Task(self, m.body['function_name'], m.body['workload'])
		task.updateInfo(m)
		task.connection_id = m.header['connection_id']
		task.status = 'working'
		callback = self.functions[task.function_name]
		
		if hasattr(callback, '__call__'):
			task.result = callback(task)
			
		return task
		
	
			
	
